### ISSUE TITLE
How can I get the extension nodes from libyang

### ISSUE THREAD

**Issue by fanchanghu** [CESNET/libyang#98]
Hi,
I can't find my extension nodes in `lys_module`, are they stored in any other place, or extension node is not supported now?
The code and schema that I used are mentioned in #95 .

**Comment by rkrejci** [comment]
You cannot, libyang completely ignores unknown extensions and skips both - the extension definitions as well as the extension instances.

**Comment by fanchanghu** [comment]
Do you have any plan to support it?

**Comment by rkrejci** [comment]
not now, this feature is problematic, since RFC is not clear how to validate its usage in schema. It is not defined which YANG statements can be used as substatements to the extension or how to apply refinement or deviation.

**Comment by fanchanghu** [comment]
I think the extension feature is just a hook in YANG, schema writer should explain how it works.  So I hope libyang can parse the extension definitions and the extension instances, and keep them in `lys_module`, then custom code can get them and explains how they work.
It means that libyang does not need to know  which YANG statements can be used as substatements to the extension or how to apply refinement or deviation. It is custom code's duty if they use the extension feature.

**Comment by rkrejci** [comment]
but the explanation how it works, how it should validate, what can be a substatement, if it is a terminating node or it can have subelements - all these things can be described in extension description and I don't see a way how to parse such a human description which can change from one to one extension. Storing extension definition is probably no problem and it can be added into libyang. 

But if you think that libyang does not need to know anything about extension's substatements, please provide an idea how to parse and store the extension instance in a schema tree. For example in case of container, libyang knows that it can contain some other nodes like leaf, list or another container and on the other hand there cannot be a type and the data instance is not supposed to contain a data value. In case of extension, libyang does not have this information. It can be described in the extension's description, but libyang is not able to get this information and therefore it is not able to check that the extension is instantiated correctly.

There is already support for some well known extensions (NACM), but it was coded into libyang and libyang knows how to handle instances of these extensions. The same can be done for other extensions, but it will probably always require some code in libyang. I'm currently not aware about a generic way of handling extension instances.

**Comment by fanchanghu** [comment]
I think the extension instance can be considered as a `lys_node_`, such as `lys_node_extinst`? When building schema tree, this node can be used as child to any `lys_node_`, such as `lys_node_container`, `lys_node_leaf`, `lys_node_rpc` or itself.

For schema tree validation, libyang just need to validate whether the extension definition exists for every extension instance, and whether the argument in extension instance is defined in extension definition.

For example, the extension definition like this:
```
	extension b{
		argument "value";
	}
```
means that the extension instance should be like this:
`	ex:b  "1111";`
in yang or like this:
`	<ex:b value="1111"/>`
in yin.

For data tree validation, libyang can ignore all the lys_node_extinst node in schema tree, and just do it as before.

**Comment by rkrejci** [comment]
One thing I don't like is that `lys_node_extinst` is going to have a space for any YANG substatement - must, typedefs, when, if-feature,..., so it is going to be the biggest structure and the most of its members will not be used. So it is going to waste the memory, but ok, I understand that it can be acceptable.

But there is another issue - what about mapping between the schema and data instances? For example `nacm:default-deny-write` and `nacm:default-deny-all` do not appear in data tree. On the other hand, `nc:get-filter-element-attributes` appear in XML data as an attribute. I can imagine another extension (even without argument specification) that is supposed to appear in data tree as a regular node. How to distinguish these situations? I think that detection that something is missing in the data tree is crucial and at least this "validation" must be done by libyang. But how libyang can know that some node is expected there? The extension instance can have `mandatory` as its substatement. What does it mean? A node must be present in data tree, an attribute must be present in data tree, or something else? It can be described in details in the extension description, but libyang know nothing about it.

**Comment by fanchanghu** [comment]
Hi,
I think the extension label will never appear in data tree, the extension just allow user to add some custom labels to schema, but the nodes in data tree can only be defined by YANG, like container, list, leaf or leaflist. 

The `lys_node_extinst` will not become too big:
First, like any other `lys_node_`, it has 4 pointers: parent, child, next and prev.
On the other hands, only the argument defined by extension definition can appear in extension instance, so any other YANG substatement - must, typedefs, when, if-feature and so on, can not appear in extension instance node.

**Comment by rkrejci** [comment]
RFC says something else:
```
...                                     The substatements of an
extension are defined by the "extension" statement, using some
mechanism outside the scope of this specification.  Syntactically,
the substatements MUST be YANG statements, or also extensions defined
using "extension" statements. 
...
An extension can allow refinement (see Section 7.13.2) and deviations
(Section 7.20.3.2),...
```
so there is no limitation for substatements. And there is no limitation for type of the statement being defined. Check [YANG metadata](https://tools.ietf.org/html/draft-ietf-netmod-yang-metadata-07) which are defined as an extension. As substatements, there are 'description', 'if-feature', 'reference', 'status', and 'units' and they appear in data tree.

**Comment by fanchanghu** [comment]
Hi,
Thanks for your explanation, you are right, the `lys_node_extinst` may become big, can we only support some commonly YANG substatements in extension instance in the first stage?
As you said, the extension instance could appear as an attribute in data tree (it seems that adding a regular node to data tree by extension is not very useful), libyang does not know how to validate it, maybe we can just keep all information in data tree and let users validate it themselves? Will additional attributes in data tree effect libyang's validation?

**Comment by rkrejci** [comment]
Ideally, I would like to have some kind of plugin mechanism for extension implementations - you could be able to prepare functions to parse the extension (in YIN and YANG formats), to define its structure in the schema tree with some properties like "how does it appear in data tree" and so on. I don't have any proposal for this since we don't need that and it seems like a lot of work, so it is just an idea. 

A simpler workaround for now could be a generic (big) structure to cover all possibilities. No validation (like which substatements are allowed and so on) would be performed in this case and the extensions would be allowed only for schemas (their instances in data tree would not be expected). But I don't have resources (and I don't want) to do that now. I would like to implement the mentioned plugins idea at the end of this or at the beginning of the next year. But as I mentioned, this is not a priority for us, so it can be done even later.

**Comment by fanchanghu** [comment]
I also think about the plugin mechanism for extension implementations, but I found that there would be 2 questions:
1, There will be a lot of repeated code when user try to parse YANG statement in extension instance, such as must, if-feature, description... but this should be done in libyang. A compromised way is that we still parse all YANG statements in libyang and keep them in a big structure, but user can prepare functions to validate them.
2, When the sysrepo project who uses libyang works in daemon mode, user can not add code to its server (it is released as an executable program), so they need another mechanism to do the validation.

**Comment by rkrejci** [comment]
ad 1 - libyang could provide some supporting functions, or it could be possible to let libyang parse well-known content of the extension instance with specifying what is allowed (or just by checking the returned subtree)

ad 2 - the extension can be provided as .so files (plugins) in some path (e.g. /usr/lib/libyang/), so the application using libyang wouldn't have anything to do with these plugins

**Comment by rkrejci** [comment]
Hi, YANG extensions are now implemented in the devel branch (versions 0.12.x), some description can be found in [documentation](https://netopeer.liberouter.org/doc/libyang/devel/howtoschemaextensions.html)
